server/fleet/integrations.go

```diff
@@ -15,120 +17,161 @@ type TeamIntegrations struct {
 	Zendesk []*TeamZendeskIntegration `json:"zendesk"`
 }
 
-// ToIntegrations converts a TeamIntegrations to an Integrations struct.
-func (ti TeamIntegrations) ToIntegrations() Integrations {
-	var intgs Integrations
-	intgs.Jira = make([]*JiraIntegration, len(ti.Jira))
-	for i, j := range ti.Jira {
-		intgs.Jira[i] = j.ToJiraIntegration()
+// MatchWithIntegrations matches the team integrations to their corresponding
+// global integrations found in globalIntgs, returning the resulting
+// integrations struct. It returns an error if any team integration does not
+// map to a global integration, but it will still return the complete list
+// of integrations that do match.
+func (ti TeamIntegrations) MatchWithIntegrations(globalIntgs Integrations) (Integrations, error) {
+	var result Integrations
+
+	jiraIntgs, err := IndexJiraIntegrations(globalIntgs.Jira)
+	if err != nil {
+		return result, err
+	}
+	zendeskIntgs, err := IndexZendeskIntegrations(globalIntgs.Zendesk)
+	if err != nil {
+		return result, err
+	}
+
+	var errs strings.Builder
+	for _, tmJira := range ti.Jira {
+		key := tmJira.UniqueKey()
+		intg, ok := jiraIntgs[key]
+		if !ok {
+			if errs.Len() > 0 {
+				errs.WriteByte('\n')
+			}
+			fmt.Fprintf(&errs, "unknown Jira integration for url %s and project key %s", tmJira.URL, tmJira.ProjectKey)
+			continue
+		}
+		intg.EnableFailingPolicies = tmJira.EnableFailingPolicies
+		result.Jira = append(result.Jira, &intg)
+	}
+	for _, tmZendesk := range ti.Zendesk {
+		key := tmZendesk.UniqueKey()
+		intg, ok := zendeskIntgs[key]
+		if !ok {
+			if errs.Len() > 0 {
+				errs.WriteByte('\n')
+			}
+			fmt.Fprintf(&errs, "unknown Zendesk integration for url %s and group ID %v", tmZendesk.URL, tmZendesk.GroupID)
+			continue
+		}
+		intg.EnableFailingPolicies = tmZendesk.EnableFailingPolicies
+		result.Zendesk = append(result.Zendesk, &intg)
 	}
-	intgs.Zendesk = make([]*ZendeskIntegration, len(ti.Zendesk))
-	for i, z := range ti.Zendesk {
-		intgs.Zendesk[i] = z.ToZendeskIntegration()
+
+	if errs.Len() > 0 {
+		err = errors.New(errs.String())
 	}
-	return intgs
+	return result, err
+}
+
+// Validate validates the team integrations for uniqueness.
+func (ti TeamIntegrations) Validate() error {
```

Comment:
this function seems to be the same as `IndexJiraIntegrations` + `IndexZendeskIntegrations`

```go
	if _, err := IndexJiraIntegrations(globalIntgs.Jira); err != nil {
		return err
	}

	if _, err := IndexZendeskIntegrations(globalIntgs.Zendesk); err != nil {
		return err
	}
```

I'm not saying this ^ is the right thing to do, but should we do something about that logic? seems to be in multiple other places too.

Reply:
Yeah the problem is that the existing functions work on `JiraIntegrations` (or zendesk) and here we process `TeamJiraIntegrations`... I had a complete set of methods on the previous implementation that converted from one to the other but they don't make sense anymore with the new changes where team integrations don't have the username/api token/etc. But I think I have an idea without forcing the use of Go1.18 just yet with generics, I'll give it a try.